Content.Server/Atmos/Portable/SpaceHeaterSystem.cs

```diff
@@ -0,0 +1,198 @@
+        thermoMachine.Cp = spaceHeater.HeatingCp;
+        thermoMachine.HeatCapacity = spaceHeater.PowerConsumption;
+    }
+
+    private void OnBeforeOpened(EntityUid uid, SpaceHeaterComponent spaceHeater, BeforeActivatableUIOpenEvent args)
+    {
+        DirtyUI(uid, spaceHeater);
+    }
+
+    private void OnUIActivationAttempt(EntityUid uid, SpaceHeaterComponent spaceHeater, ActivatableUIOpenAttemptEvent args)
+    {
+        if (!Comp<TransformComponent>(uid).Anchored)
+        {
+            _popup.PopupEntity(Loc.GetString("comp-space-heater-unanchored", ("device", Loc.GetString("comp-space-heater-device-name"))), uid, args.User);
+            args.Cancel();
+        }
+    }
+
+    private void OnDeviceUpdated(EntityUid uid, SpaceHeaterComponent spaceHeater, ref AtmosDeviceUpdateEvent args)
+    {
+        if (!_power.IsPowered(uid)
+            || !TryComp<GasThermoMachineComponent>(uid, out var thermoMachine))
+        {
+            return;
+        }
+
+        // First get the heat direction of the thermomachine (if any) and update appeareance accordingly
+        if (thermoMachine.LastEnergyDelta > 0)
+        {
+            spaceHeater.State = SpaceHeaterState.Heating;
+        }
+        else if (thermoMachine.LastEnergyDelta < 0)
+        {
+            spaceHeater.State = SpaceHeaterState.Cooling;
+        }
+        else
+        {
+            spaceHeater.State = SpaceHeaterState.StandBy;
+        }
+
+        UpdateAppearance(uid, spaceHeater.State);
+
+        // Then, if in automatic temperature mode, check if we need to adjust the heat exchange direction
+        if (spaceHeater.Mode == SpaceHeaterMode.Auto)
+        {
+            var environment = _atmosphereSystem.GetContainingMixture(uid);
+            if (environment == null)
+                return;
+
+            if (environment.Temperature <= thermoMachine.TargetTemperature - (thermoMachine.TemperatureTolerance + spaceHeater.AutoModeSwitchThreshold))
+            {
+                thermoMachine.Cp = spaceHeater.HeatingCp;
+            }
+            else if (environment.Temperature >= thermoMachine.TargetTemperature + (thermoMachine.TemperatureTolerance + spaceHeater.AutoModeSwitchThreshold))
+            {
+                thermoMachine.Cp = spaceHeater.CoolingCp;
+            }
+        }
+    }
+
+    private void OnPowerChanged(EntityUid uid, SpaceHeaterComponent spaceHeater, ref PowerChangedEvent args)
+    {
+        UpdateAppearance(uid, spaceHeater.State);
+        DirtyUI(uid, spaceHeater);
+    }
+
+    private void OnToggle(EntityUid uid, SpaceHeaterComponent spaceHeater, SpaceHeaterToggleMessage args)
+    {
+        ApcPowerReceiverComponent? powerReceiver = null;
+        if (!Resolve(uid, ref powerReceiver))
+            return;
+
+        _power.TogglePower(uid);
+        if (powerReceiver.PowerDisabled)
+            spaceHeater.State = SpaceHeaterState.Off;
+        else
+            spaceHeater.State = SpaceHeaterState.StandBy;
+
```

Comment:
For example, doesn't this duplicate a check that is made immediately in `UpdateAppearance()`?